### PR TITLE
Prepare DB changes for moving to immutable transactions

### PR DESCRIPTION
**Pull request by chime-mu**
Finally it's finished. It's been quite a lot of work and become a huge change. I know it's hard to review these really big changes, so I'm sorry about that.

Below is a few hints to make testing easier. 

# To test this

* Download the latest production database dump from the S3 bucket `database-dumps.cobudget.co` 
* If needed, change the Owner in the DB dump using sed:
	*  `sed -e 's/Owner: cobudget/Owner: <your-db owner>/' -e 's/OWNER TO cobudget/OWNER TO <your-db-owner>/' cobudget-prod-171122 >db_loadable`
* Reload the database with the newest production DB
	* `dropdb cobudget_development`
	* `createdb cobudget_development`
	* `psql -d cobudget_development <db_loadable`
* Migrate `rake db:migrate`
* Now you can check that all buckets and all memberships are correct after the migration (from the rails console):
	* `BucketService.check_all_buckets`
	* `MembershipService.check_all_memberships`
* Check any other things you feel is necessary

# Other comments

* Funds added to buckets that are funding or funded will no longer be removed when a member is archived, so a small change in the front end is needed as well.
* I'm happy to go through all of this in phone call if that's helpful - or help in any other way.
* Another way to do this, would be for you to point out whatever is unclear, which would give me an opportunity to make the code more readable and possibly add more comments @jlopker

### CODE REVIEW

**Comment by jlopker** [comment]
`group_data` in `group_analytics_service.rb` needs to be updated to use the transactions table

**Comment by chime-mu** [comment]
@jlopker Agree it would be good if `group_data` used the transactions table. I figured we could wait with that change until this thing was deployed, since it's already huge.

In general the `transactions` table is only updated but never actually read from (yet). The next steps would be to move towards using the `transactions` table for actual reads, deploy in smaller pieces until `allocations` and `contributions` are no longer used.

I agree ´group_data` should be one of the first things moved.

**Comment by jlopker** [comment]
@chime-mu sounds good re: `group_data`

this is what i'm seeing- when new money gets added to someone's account, the `from_account` and the `to_account` is linked to the same person. is that the intended behavior? in a theory, it would make sense that that money would come from some "group" account, but that doesn't exist, so i'm guessing this was what you thought made most sense. is that correct?

File: app/mailers/user_mailer.rb [hunk 1]
```diff
@@ -81,4 +81,23 @@ def recent_activity_digest_email(user:)
       )
     end
   end
+
+  def notify_admins_funds_is_returned_to_group_account(admin:, bucket:, done_by:, archived_member:, amount:, group_account:)
```
Comment (jlopker) [inline app/mailers/user_mailer.rb]: should rename to `notify_admins_funds_are_returned_to_group_account`

File: app/mailers/user_mailer.rb [hunk 2]
```diff
@@ -81,4 +81,23 @@ def recent_activity_digest_email(user:)
...
+    @bucket = bucket
+    @group = bucket.group
+    @done_by = done_by
+    @archived_member = archived_member
+    @amount = amount
+    @formatted_amount = Money.new(amount * 100, @group.currency_code).format
+    @group_account = group_account
+    mail(to: admin.name_and_email,
+         from: "Cobudget Updates <updates@cobudget.co>",
+         subject: "Funds from cancelled bucket is returned to group account")
```
Comment (jlopker) [inline app/mailers/user_mailer.rb]: should be - "Funds from cancelled bucket have been returned to group account"

File: app/models/group.rb [hunk 3]
```diff
@@ -47,6 +47,35 @@ def extend_trial
...
   end
 
+  def ensure_group_user_exist()
+    uid = %(group@group-#{id}.co)
+    group_user = User.find_by uid: uid
+    if !group_user 
+      group_user = User.create!({
+          name: %(Group "#{name}"),
+          uid: uid,
+          email: uid,
```
Comment (jlopker) [inline app/models/group.rb]: email ends up being something like "group@group-315.co" 
I feel like this is a little bit strange bc this looks like a real email address but isn't..

File: app/services/bucket_service.rb [hunk 4]
```diff
@@ -1,12 +1,61 @@
...
+    ActiveRecord::Base.transaction do
+      if bucket.is_idea? || bucket.is_funding? || bucket.is_funded?
+        bucket.contributors(exclude_author: exclude_author_from_email_notifications).each do |funder|
+          # Don't notify archived members
+          if !Membership.find_by(group_id: bucket.group_id, member_id: funder.id).archived_at
+            UserMailer.notify_funder_that_bucket_was_archived(funder: funder, bucket: bucket).deliver_later
+          end
+        end
+        bucket.update(archived_at: DateTime.now.utc)
+        bucket.contributions.destroy_all
```
Comment (jlopker) [inline app/services/bucket_service.rb]: have you thought about not destroying the contributions and instead archiving them? do you think this history interesting and/or useful to preserve?

File: app/views/user_mailer/notify_admins_funds_is_returned_to_group_account.erb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+<p>
+  A bucket '<%= @bucket.name %>' was cancelled in <%= @group.name %> 
+  by <%= @done_by.name %>.
+</p>
+
+<p>
+  A previous member <%= @archived_member.name %>, had funded the bucket with 
+  <%= @formatted_amount %>. As this person is no longer a member of the group, the funds
+  has been returned to the group account <%= @group_account.name %>.
```
Comment (jlopker) [inline app/views/user_mailer/notify_admins_funds_is_returned_to_group_account.erb]: should be "have been"

File: app/mailers/user_mailer.rb (hunk 1 above)
Comment (chime-mu) [reply app/mailers/user_mailer.rb]: Oh - thanks. Done.

File: app/mailers/user_mailer.rb (hunk 2 above)
Comment (chime-mu) [reply app/mailers/user_mailer.rb]: Done.

File: app/models/group.rb (hunk 3 above)
Comment (chime-mu) [reply app/models/group.rb]: Fair point. The reason is the User model is quite restrictive with the shape of Emails. It actually has to look like a real email address. We could of course loosen that restriction or move the validation to only be relevant for emails entered by users.

I get that the idea of using existing users and memberships for the group account is kind of cheating and it feels a bit like forcing different concepts into a single representation.

What would be your preferred solution here?

File: app/services/bucket_service.rb (hunk 4 above)
Comment (chime-mu) [reply app/services/bucket_service.rb]: Yes, I think it's essential we preserve the history. This is one of the main changes behind this whole thing. As you can see, the transaction records are not being destroyed. Instead new "reverse" transactions are entered, so we never loose history.

The idea is to eventually get rid of the `contributions` table and the `allocations` table completely so both are completely replaced by the information in the `transactions` table. This should already be the case, as all updates should be reflected in the `transactions`. You will also notice I've created a job to be run daily to check that the old `contributions` and `allocations` system is in alignment with the `tranactions` system.

In reality what I'm trying to do here is to release a change in steps - before it's fully finished. This step is updating all the functions that write money information to update both "systems". Next step will be to change the different functions that read and return data to use the new `transactions` system as well.

Does this make sense?

File: app/views/user_mailer/notify_admins_funds_is_returned_to_group_account.erb (hunk 5 above)
Comment (chime-mu) [reply app/views/user_mailer/notify_admins_funds_is_returned_to_group_account.erb]: Done

**Comment by chime-mu** [comment]
@jlopker Couldn't sleep so now I'm doing this instead :)

Thank you for checking the language - much appreciated!

You asked
> when new money gets added to someone's account, the `from_account` and the `to_account` is linked to the same person. is that the intended behavior? in a theory, it would make sense that that money would come from some "group" account, but that doesn't exist, so i'm guessing this was what you thought made most sense. is that correct?

Yes, this is the intended behaviour, but it's not because it's correct in any objetive way. It's a decision.

The way it's done right now, all money coming in to cobudget is accounted in different accounts, one for each user. I assumed that money was coming in from individual people. In the Enspiral Services and Outlandish case money is added dependent on how much money people have earned. For Ouishare I think it is money directly paid by the user and the same goes for Enspiral Foundation. If you want to see the total sum of money paid by a user, the is that users incoming account. If you want to see the sum of everything paid to the group it's the sum of money paid by all users. If your want to see the sum of money a user has received through buckets, it's the outgoing account. If you want to see the net balance of money paid and received it's incoming minus outgoing.

We could say all money could come from some group account as you suggest, but then first of all I don't think it reflects the real world reality - and second, it would only give a "total paid in for the group" sum and not for individual users.

When I've used a group account for money from archived members and when money should be returned to archived members, the word "group account" is actually not very precise. It's more like a "We cant decide automatically in the system what to do with this money, so we'll put it here so you admins can make a decision" account. Some may chose to pay it back to the original funders (if the money in cobudget is in reality only being held by organisation using cobudget), others might choose to redistribute (if the money in cobudget is actually owned by the org using cobudget). Some may use it to buy ice cream :)

I hope this makes sense. Let me know if it still feels unclear.

File: app/models/group.rb (hunk 3 above)
Comment (jlopker) [reply app/models/group.rb]: hmmm how about we do group-315@cobudget.co ?

File: app/services/bucket_service.rb (hunk 4 above)
Comment (jlopker) [reply app/services/bucket_service.rb]: ah yes of course that makes sense! thx for the clarification

**Pull request merged**